Add rukpakctl subcommands that utilize local source bundle capabilities.

akihikokuroda opened the pull request:
This PR has primitive subcommands for development use.

rukpakctl create bundle <bundle name> <manifest directory> - create bundle with the contents of the manifest directory
```
bin/rukpakctl create bundle test1 testdata/bundles/plain-v0/valid/manifests
```
rukpakctl create bundledeployment <bundledeployment name> <manifest directory> - create bundledeployment with the contents of the manifest directory
```
bin/rukpakctl create bundledeployment test2 testdata/bundles/plain-v0/valid/manifests
```
rukpakctl content <bundle name> - show the contents of the bundle
```
bin/rukpakctl content test1
```
Signed-off-by: akihikokuroda <akuroda@us.ibm.com>

exdx commented on cmd/rukpakctl/cmd/bundle.go:
```diff
@@ -0,0 +1,113 @@
+
+func init() {
+	createCmd.AddCommand(bundleCmd)
+	homeDir := os.Getenv("HOME")
+	config, err := clientcmd.BuildConfigFromFlags("", homeDir+"/.kube/config")
```
I'm not sure this is the best way to do this. I believe kubectl has an exported function to get config from the machine, could we use that instead? 

If we do rely on a `HOME` env variable then we would want to document it

akihikokuroda replied:
Thanks for review.  I'll look into it.

akihikokuroda replied:
I fixed it.

exdx commented on cmd/rukpakctl/cmd/content.go:
```diff
@@ -0,0 +1,188 @@
+
+var contentCmd = &cobra.Command{
+	Use:   "content <bundle name>",
+	Short: "content command display contents of the specified bundle.",
+	Long:  `content command display contents of the specified bundle.`,
```
nit: I don't think we need to repeat the name of the command in the help description

exdx commented on cmd/rukpakctl/cmd/content.go:
```diff
@@ -0,0 +1,188 @@
+	// Create a temporary ServiceAccount
+	sa := corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "rukpakctl-sa",
```
nit: I think we should specify the rukpak-system namespace here

exdx commented on cmd/rukpakctl/cmd/content.go:
```diff
@@ -0,0 +1,188 @@
+	if err = opt.k8sclient.CoreV1().ServiceAccounts("default").Delete(context.Background(), "rukpakctl-sa", metav1.DeleteOptions{}); err != nil {
+		return fmt.Errorf("failed to delete service account: %+v", err)
+	}
+	if err = opt.k8sclient.RbacV1().ClusterRoleBindings().Delete(context.Background(), "rukpakctl-crb", metav1.DeleteOptions{}); err != nil {
+		return fmt.Errorf("failed to delete clusterrolebinding: %+v", err)
+	}
+	if err = opt.k8sclient.BatchV1().Jobs("default").Delete(context.Background(), "rukpakctl-job", metav1.DeleteOptions{PropagationPolicy: &option}); err != nil {
+		return fmt.Errorf("failed to delete job: %+v", err)
+	}
+	if err = opt.k8sclient.CoreV1().Pods("default").Delete(context.Background(), pods.Items[0].Name, metav1.DeleteOptions{}); err != nil {
+		return fmt.Errorf("failed to delete pod: %+v", err)
```
alternatively, we could create the cluster role binding first, and then ownerref everything by it, so all these objects go away automatically. 

I believe creating a cluster role binding pointing to a non-existent SA is fine, on a further reconcile it will find the SA and succeed

akihikokuroda replied:
It's a good way for clean up.  I'll try it.

akihikokuroda replied:
It works well.  Tnanks!

exdx reviewed (commented):
Nice job! I think this is a great start to making `rukpakctl` a useful tool.

akihikokuroda commented on cmd/rukpakctl/cmd/bundle.go:
```diff
@@ -0,0 +1,113 @@
+				Local: &rukpakv1alpha1.LocalSource{
+					ConfigMapRef: &rukpakv1alpha1.ConfigMapRef{
+						Name:      configmapName,
+						Namespace: "default",
```
OK. I'll change it.

akihikokuroda commented on cmd/rukpakctl/cmd/bundledeployment.go:
```diff
@@ -0,0 +1,119 @@
+			ProvisionerClassName: "core.rukpak.io/plain",
+			Template: &rukpakv1alpha1.BundleTemplate{
+				Spec: rukpakv1alpha1.BundleSpec{
+					ProvisionerClassName: "core.rukpak.io/plain",
```
It creates the local source bundle for now in this subcommand. So it only supports plain.  We can enhance this to support the other provisioner later.

exdx replied:
Ok, fair enough

akihikokuroda commented on cmd/rukpakctl/utils/utils.go:
```diff
@@ -0,0 +1,47 @@
+	configmap := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: name,
+			Namespace:    "default",
```
I'll make change.  Thanks!

akihikokuroda commented on cmd/rukpakctl/cmd/bundle.go:
```diff
@@ -0,0 +1,113 @@
+		return fmt.Errorf("failed to convert to unstructured: %+v", err)
+	}
+	bundleGVR := schema.GroupVersionResource{Group: "core.rukpak.io", Version: "v1alpha1", Resource: "bundles"}
+	_, err = opt.k8sdynamic.Resource(bundleGVR).Create(context.Background(), &unstructured.Unstructured{Object: bundleUnstructured}, metav1.CreateOptions{})
```
I used to use dynamic client for CRD instances.  I'll change them to the typed client.  Thanks!

akihikokuroda commented on cmd/rukpakctl/cmd/content.go:
```diff
@@ -0,0 +1,188 @@
+	}
+
+	// Get Pod for the Job
+	pods, err := opt.k8sclient.CoreV1().Pods("default").List(context.Background(), metav1.ListOptions{LabelSelector: "job-name=rukpakctl-job"})
```
The label is automatically added by the Job when the pod is created.  I change the name of the pod to generatename so the pod name for the job will be unique.

akihikokuroda commented:
I addressed all comments.  They have `kubeconfig` for the kubernetes cluster access config and `namespace` for the nemespace of the target or work resources namespace flags now.

exdx commented on cmd/rukpakctl/cmd/bundledeployment.go:
```diff
@@ -0,0 +1,140 @@
+	Long:  `create rukpak bundledeployment resource with specified contents.`,
+	Args: func(cmd *cobra.Command, args []string) error {
+		if len(args) < 2 {
+			return errors.New("requires 2 arguments: <bundle name>, <manifest file directory>")
```
Same as above

akihikokuroda replied:
I'll update.

exdx commented on cmd/rukpakctl/cmd/bundledeployment.go:
```diff
@@ -0,0 +1,140 @@
+			APIVersion: "core.rukpak.io/v1alpha1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: args[0],
```
Same as above

akihikokuroda replied:
I'll update.

exdx commented on cmd/rukpakctl/cmd/content.go:
```diff
@@ -0,0 +1,219 @@
+			return
+		}
+
+		scheme := runtime.NewScheme()
+		err = rukpakv1alpha1.AddToScheme(scheme)
+		if err != nil {
+			fmt.Printf("failed to add schema: %+v\n", err)
+			return
```
I wonder, could we the scheme in an init somewhere at the root-level of the tool? So we don't have to add to the scheme in each command separately.

akihikokuroda replied:
I'll move it to the rootCmd.

exdx commented on cmd/rukpakctl/cmd/content.go:
```diff
@@ -0,0 +1,219 @@
+	}
+	url := bundle.Status.ContentURL
+
+	// Create a Job that reada from the URL and outputs contents in the pod log
```
```suggestion
	// Create a Job that reads from the URL and outputs contents in the pod log
```

exdx commented on cmd/rukpakctl/cmd/create.go:
```diff
@@ -0,0 +1,33 @@
+)
+
+// createCmd represents the create command
+var createCmd = &cobra.Command{
```
can we remove this if it's not going to be defined in this PR?

exdx reviewed (commented):
Looks good. Is the plan to add some tests as a follow-up?

akihikokuroda commented on cmd/rukpakctl/cmd/bundle.go:
```diff
@@ -0,0 +1,135 @@
+)
+
+type BundleOptions struct {
+	k8sclient *kubernetes.Clientset
+	rtclient  runtimeclient.Client
```
OK. I'll try.

akihikokuroda commented on cmd/rukpakctl/cmd/bundle.go:
```diff
@@ -0,0 +1,135 @@
+			APIVersion: "core.rukpak.io/v1alpha1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: args[0],
```
I'll update.

akihikokuroda commented on cmd/rukpakctl/cmd/bundledeployment.go:
```diff
@@ -0,0 +1,140 @@
+
+type BundleDeploymentOptions struct {
+	k8sclient *kubernetes.Clientset
+	rtclient  runtimeclient.Client
```
I'll update.

akihikokuroda commented:
All comments were addressed.  The comment about `createCmd` needs a response.  Thanks!

akihikokuroda commented:
I'll create an issue for test for this later.
